### ISSUE TITLE
Loading product page always write something to DB

### ISSUE THREAD

**Issue by erikdemarco** [woocommerce/woocommerce#30349]
Writing to database every pageload is very bad idea (especially on frontend)

To replicate:
-) Install Query Monitor
-) Load product page
-) see list of 'UPDATE' queries from QM  console

Please read Wordpress code quality & best pratice here:
=> https://docs.wpvip.com/technical-references/code-review/vip-warnings/#h-front-end-database-writes
=> https://docs.wpvip.com/technical-references/code-quality-and-best-practices/term-queries-should-consider-include_children-false/

The problem is in wc_get_related_products() function here:
https://github.com/woocommerce/woocommerce/blob/b19500728b4b292562afb65eb3a0c0f50d5859de/includes/wc-product-functions.php#L894

If the related products count is under limit it will keeps doing this expensive taxonomy queries every single pageload. and set_transient is the one which will do 'UPDATE' query to db to update the transient_timeout

Proposed solution:
-) If related products count is under limit we just need to lower the timeout, from the default 'DAY_IN_SECONDS' to 'MINUTE_IN_SECONDS' or 'HOUR_IN_SECONDS'

**Comment by DavidAnderson684** [comment]
N.B.  WP VIP deploys an object cache, so won't see any database write here: https://docs.wpvip.com/technical-references/caching/object-cache/ .

**Comment by vedanshujain** [comment]
Agree with the suggestion, we should be busting the transient on the product save hook, not using it when we have product below limits does not make much sense.
So to fix, we should:
1. Remove the count check.
2. Bust the transient on product save hook.

**Comment by erikdemarco** [comment]
@DavidAnderson684 Its unrelated to object cache. Its just a good code practice.

But even if very high traffic site using object cache. It will trigger another bug. making object cache sometimes out of sync because of race condition.:
=> https://core.trac.wordpress.org/ticket/31245

This quote exactly describe this case:
(We're running into this issue a few times a week, due to certain plugins running an update_option on a high percentage of requests. These plugins are obviously being dumb in doing so, but they're revealing the core bug quite nicely.)

**Comment by Patil2099** [comment]
I would love to take this up

**Comment by cpapazoglou** [comment]
> Agree with the suggestion, we should be busting the transient on the product save hook, not using it when we have product below limits does not make much sense. So to fix, we should:
> 
> 1. Remove the count check.
> 2. Bust the transient on product save hook.

Indeed, the `count` check seems redundant. Without it, we need a strategy to clear the transient. I think that busting the transient on product save hook will not work as expected. Currently, the feature works by saving related product Ids in a transient for the current product - let's call it `master product`. So, each product has a `'_transient_timeout_wc_related_' . $product_id` transient with a timestamp value as expiration date.

- ✅ Saving the `master product` will query for related products and set the transient for the `master product`. Visiting the master product will show the correct related products.
- ❌ Saving any `other product` so that it gets related to `master product` will query for related products and set the transient for that `other product`. Visiting this product will show the correct related products but visiting the `master product` will not show this `other product` due to the transient of the `master product` not being busted.

**Ideas:**
- [Best Performance - Worst UX] It's OK to show stale related products. If the product doesn't exist it won't be rendered anyway. Drawback: if a related product was wrongly set ( example wrong category ), it will take 1 day to get removed.
- [Worst Performance] On product save:
  - Get the related products of the current product and set the transient
  - For each related product, update their transient ( expensive )
- [Good Performance] Add a new _maintenance_ feature that clears all `_transient_timeout_wc_related_%`. We can educate the user that for performance reasons related products are cached for 1 day and if needed they can manually bust the cache.
**- [OK Performance - Best UX] On product save, we can clear all `_transient_timeout_wc_related_%`. Deleting multiple DB records should be cheap and they will get populated again one-by-one when users visit a page that queries for related products ( once per day ).** 

The last suggestion is followed in https://github.com/woocommerce/woocommerce/pull/55843

**Comment by cpapazoglou** [comment]
Actually, just found that the `master product` transient is already deleted when saving the product in https://github.com/woocommerce/woocommerce/blob/39239cff88f96776b2a1f2d4da15f3fa9ceebf25/plugins/woocommerce/includes/wc-product-functions.php#L145

**Comment by cpapazoglou** [comment]
Switched to an approach that clears **only affected related products** in an async action to avoid the performance penalty of `wc_get_related_products( $post_id, 1000 );` in product-> save().
